Add json output format to check

martinemde opened the pull request:
This change adds a json output option and provides a schema to describe the output format. The goal is to be able to programmatically parse the output of `pks check` for other tools like danger.

We decided to split the idea of a stale todo from the idea of a violation, since even though they talk about the same idea, the have different meanings.

- A stale todo is a leftover reference to a problem that no longer exists. The fix requires no work other than running `pks update`. It's a lint warning.
- A violation should be address via a choice by the developer to either repair the violation, change the package config, or add it to the todo.

martinemde committed:
Implement json output format option

Validate against the included schema for check json output

Co-authored-by: Ivy Evans <ivy.evans@gusto.com>

martinemde committed:
Add line and column to json output for violations

ivy commented on src/packs/json.rs:
```diff
@@ -0,0 +1,104 @@
+use itertools::chain;
```
```suggestion
//! JSON output formatter for `pks check -o json`.
//!
//! Serializes check results (violations, stale TODOs, and summary) to JSON.
//! See `schema/check-output.json` for the JSON Schema specification.

use itertools::chain;
```

ivy commented on tests/check_test.rs:
```diff
@@ -399,3 +419,194 @@ fn test_check_contents_ignoring_recorded_violations(
+        .arg("-o")
+        .arg("json")
+        .assert()
+        .success()
```
```suggestion
        .failure()
```

ivy commented on tests/check_test.rs:
```diff
@@ -399,3 +419,194 @@ fn test_check_contents_ignoring_recorded_violations(
+        .arg("-o")
+        .arg("json")
+        .assert()
+        .success()
```
```suggestion
        .failure()
```

ivy commented on src/packs.rs:
```diff
@@ -86,6 +87,9 @@ pub fn check(
             csv::write_csv(&result, std::io::stdout())?;
         }
+        OutputFormat::JSON => {
+            json::write_json(&result, std::io::stdout())?;
```
We should exit with a non-zero status when there are violations:

```suggestion
            json::write_json(&result, std::io::stdout())?;
            if result.has_violations() {
                std::process::exit(1);
            }
```

I wonder if we should use a different code versus a runtime error? This way, callers don't have to differentiate "runtime error as plain text" versus "violations presented as JSON".

ivy reviewed (changes requested):
Looks good to me outside of a design detail: I think we should exit with a non-zero status code when there are violations.

Also suggested a doc comment to point readers/agents to the JSON schema. Love that you added a test that validates against it! 💜

martinemde committed:
Control exit status for `check` command

This follows the convention used by linters like eslint, rubocop, and shellcheck.
Builds on clap's existing behavior.

All output formats (packwerk, json, csv) behave consistently:
- Exit 0 when no violations
- Exit 1 when violations found
- Exit 2 on internal errors

martinemde committed:
Update src/packs/json.rs

Co-authored-by: Ivy Evans <ivy@ivyevans.net>

martinemde committed:
cargo fmt

martinemde committed:
Don't be too preshus about exit 2 approach

martinemde committed:
Fix clippy again, for everything

hellosweta reviewed (approved):
🏄
